### PR TITLE
Fix two auto-atomicity Regex bugs

### LINKED ISSUE #74686: Regex auto-atomicity too aggressive about \w and \b
### Description

`System.Text.RegularExpressions.Regex.Matches` not returning expected matches. It returns no matches at all.

### Reproduction Steps

```csharp
var stringToMatch = "sydney bogota berlin tokyo nairobi denver rio";
Console.WriteLine($"Input: {stringToMatch}");

GetMatches(@"(\b(?!bogo|nai)\w*\b)\w+", stringToMatch);
GetMatches(@"\w*\b\w+", stringToMatch);

void GetMatches(string expression, string stringToMatch)
{
    var reg = new System.Text.RegularExpressions.Regex(expression);

    Console.WriteLine($"\nOutput:");
    foreach (var match in reg.Matches(stringToMatch))
    {
        Console.WriteLine(match);
    }
}
```

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFrameworks>net6.0</TargetFrameworks>
    <ImplicitUsings>enable</ImplicitUsings>
    <LangVersion>latest</LangVersion>
  </PropertyGroup>
</Project>
```

### Expected behavior

```
Input: sydney bogota berlin tokyo nairobi denver rio

Output:
sydney
berlin
tokyo
denver
rio

Output:
sydney
bogota
berlin
tokyo
nairobi
denver
rio
```

Example 1: https://regex101.com/r/Xn7X3w/2
Example 2: https://regex101.com/r/Q10441/1

### Actual behavior

```
Input: sydney bogota berlin tokyo nairobi denver rio

Output:

Output:
```

### Regression?

When running the same code pointing to `netstandard2.0` and `net472` it's returning correct matches. 

Whereas `.net5` and `.net6.0` returns no match at all.

### Known Workarounds

_None_

### Configuration

* .NET 6.0
* Windows 11
* x64

### Other information

Other community members in Stackoverflow were also able to reproduce this issue: https://stackoverflow.com/questions/73498919/regex-match-differs-between-net-versions

### LINKED ISSUE #74722: Incorrect behaviour of regex atomic groups
### Description

In some scenarios, the regex engine can backtrack in an atomic group. This should not be possible

### Reproduction Steps

`var m = System.Text.RegularExpressions.Regex.Match("aaa", "^(?>a*)a$")`

### Expected behavior

`m.Success == false`

The above regex should not match. The atomic group should greedily match "aaa", and then be unable to backtrack to match "aa"; and so the overall match should fail.

### Actual behavior

`m.Success == true`

The regex matches.

### Regression?

Behaviour is correct on:

.NET Core 3.1.28
.NET Framework 4.8.4515.0

Behaviour is incorrect on:

.NET 5.0.17
.NET 6.0.8
.NET 7.0.0-preview.7.22375.6

(Version names from System.Runtime.InteropServices.RuntimeInformation.FrameworkDescription)

### Known Workarounds

_No response_

### Configuration

Bug exists on
.NET 5.0.17
.NET 6.0.8
.NET 7.0.0-preview.7.22375.6

Windows 10 21H2
x64

### Other information

_No response_

### PR DESCRIPTION
**Pull request by stephentoub**
We walk concatenations in order to combine adjacent loops, e.g. `a+a+a+` becomes `a{3,}`.  We also combine loops with individual items that are compatible, e.g. `a+ab` becomes `a{2,}b`. However, we're doing these operations on atomic loops as well, which is sometimes wrong. Since an atomic loop consumes as much as possible and never gives anything back, combining it with a subsequent loop will end up essentially ignoring any minimum specified in the latter loop. We thus can't combine atomic loops if the second loop has a minimum; this includes the case where the second "loop" is just an individual item.
Fixes https://github.com/dotnet/runtime/issues/74686

We currently consider \w and \b non-overlapping, which allows a \w loop followed by a \b to be made atomic.  The problem with this is that \b is zero-width, and it could be followed by something that does overlap with the \w. When matching at a location that is a word boundary, it is possible the first loop could give up something that matches the subsequent construct, and thus it can't be made atomic. (We could probably restrict this further to still allow atomicity when the first loop has a non-0 lower bound, but it doesn't appear to be worth the complication.)
Fixes https://github.com/dotnet/runtime/issues/74722

We should backport these correctness fixes to both release/7.0 and release/6.0.

### CODE REVIEW

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs [hunk 1]
```diff
@@ -149,6 +149,15 @@ private static IEnumerable<(string Pattern, string Input, RegexOptions Options,
                     yield return (Case("(?>[^z]+)z"), "zzzzxyxyxyz123", options, 4, 9, true, "xyxyxyz");
                     yield return (Case("(?>(?>[^z]+))z"), "zzzzxyxyxyz123", options, 4, 9, true, "xyxyxyz");
                     yield return (Case("(?>[^z]*)z123"), "zzzzxyxyxyz123", options, 4, 10, true, "xyxyxyz123");
+                    yield return (Case("(?>a*)a"), "aaa", options, 0, 3, false, "");
+                    yield return (Case("(?>a*)a+"), "aaa", options, 0, 3, false, "");
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs:153]:
```suggestion
                    yield return (Case("(?>a*)a+"), "aaa", options, 0, 3, false, "");
                    yield return (Case("(?>a*)a{1,}"), "aaa", options, 0, 3, false, "");
```
etc ... or are these equivalent?

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs [hunk 2]
```diff
@@ -2103,14 +2114,14 @@ private static bool CanBeMadeAtomic(RegexNode node, RegexNode subsequent, bool i
...
-                            case RegexNodeKind.Boundary when node.Str is RegexCharClass.WordClass or RegexCharClass.DigitClass:
-                            case RegexNodeKind.NonBoundary when node.Str is RegexCharClass.NotWordClass or RegexCharClass.NotDigitClass:
-                            case RegexNodeKind.ECMABoundary when node.Str is RegexCharClass.ECMAWordClass or RegexCharClass.ECMADigitClass:
-                            case RegexNodeKind.NonECMABoundary when node.Str is RegexCharClass.NotECMAWordClass or RegexCharClass.NotDigitClass:
                                 return true;
 
                             case RegexNodeKind.Onelazy or RegexNodeKind.Oneloop or RegexNodeKind.Oneloopatomic when subsequent.M == 0 && !RegexCharClass.CharInClass(subsequent.Ch, node.Str!):
                             case RegexNodeKind.Setlazy or RegexNodeKind.Setloop or RegexNodeKind.Setloopatomic when subsequent.M == 0 && !RegexCharClass.MayOverlap(node.Str!, subsequent.Str!):
+                            case RegexNodeKind.Boundary when node.Str is RegexCharClass.WordClass or RegexCharClass.DigitClass:
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:2121]: What about \u200c and \u200d ? Those aren't RegexCharClass.WordClass or RegexCharClass.DigitClass, but qualify as a boundary, I think?

File: src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs (hunk 1 above)
Comment (danmoseley) [reply src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs:153]: I guess they are and that's protected by PatternsReduceIdentically tests.

Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/tests/FunctionalTests/Regex.Match.Tests.cs:153]: Correct

File: src/libraries/System.Text.RegularExpressions/tests/UnitTests/RegexReductionTests.cs [hunk 3]
```diff
@@ -470,6 +470,10 @@ public void PatternsReduceIdentically(string actual, string expected)
         [InlineData("(?:ab??){2}", "(?:a(?>b??)){2}")]
         [InlineData("(?:ab??){2, 3}", "(?:a(?>b??)){2, 3}")]
         [InlineData("ab??(b)", "a(?>b??)(b)")]
+        [InlineData(@"\w+\b\w+", @"(?>\w+)\b\w")]
```
Comment (danmoseley) [inline src/libraries/System.Text.RegularExpressions/tests/UnitTests/RegexReductionTests.cs:473]: I'm confused, isn't this reduction what the bug fix is supposed to block? 

edit, nuts, again I'm confused by Github showing the method name above, but this applies to PatternsReduceDifferently

**Review by danmoseley — APPROVED** [review]
@joperezr should review of course, as perhaps he can think of other variations

File: src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs (hunk 2 above)
Comment (stephentoub) [reply src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexNode.cs:2121]: They do, but it shouldn't matter. The question to ask is, "is there anything the loop could match that could also be matched by what comes after it?" And \w couldn't match those additional characters.

**Comment by stephentoub** [comment]
> (We could probably restrict this further to still allow atomicity when the first loop has a non-0 lower bound, but it doesn't appear to be worth the complication.)

For context, I tried this PR out on the ~19K regexes in our corpus. A grand total of 9 were impacted by the \w\b fix (meaning 9 of the patterns had a loop that was being made atomic that with this fix is no longer being made atomic). When I changed that to only consider a non-0 lower bound, the 9 only dropped to 8.

**Review by joperezr — APPROVED** [review]
Changes look good to me, thanks for fixing and thanks to @mathy-plutoflume and @Mankarse for the reports!

Just curious, did you happen to run the perf tests after the change? I believe I saw some cases that might be impacted so I was just wondering how much does it affect not doing auto-atomicity for those cases. Of course, correctness is more important than perf 😄

**Comment by stephentoub** [comment]
> did you happen to run the perf tests after the change? I believe I saw some cases that might be impacted

I didn't.  Can you point to which you saw?  I see one's that do `\b(\w+)\b`, but those shouldn't be negatively impacted.  I just added two more test cases to validate that.

**Comment by joperezr** [comment]
I saw this one but I guess it's not exactly the same: https://github.com/dotnet/performance/blob/e18fa8744878207f770c54295b03bbdd38b71168/src/benchmarks/micro/libraries/System.Text.RegularExpressions/Perf.Regex.Industry.cs#L165

I also recall one test I saw which you wrote, (but not sure if it was a perf test) where for a given input you would create Regexes for each word found, and then count how many times that given word showed up in the text. I can't find that one but that was one that I thought might be something that could be impacted by this.

**Comment by stephentoub** [comment]
> I saw this one but I guess it's not exactly the same:

Yeah, that one has `\w+n`, so it wouldn't have been made atomic anyway, since `n` is part of `\w`.

> where for a given input you would create Regexes for each word found

Yup, that's here:
https://github.com/dotnet/performance/blob/e18fa8744878207f770c54295b03bbdd38b71168/src/benchmarks/micro/libraries/System.Text.RegularExpressions/Perf.Regex.Industry.cs#L108
That's using `@"\b(\w+)\b"`, in which that loop will still be made atomic.

**Comment by craig-plutoflume** [comment]
@stephentoub - Mathy from my company found this during our upgrade from .NetCore 3.1 to .Net 6.0.  Is there any chance this could also get ported to `release/6.0`?  Without this fix we're blocked for migration and will have to figure something else out.

**Comment by stephentoub** [comment]
@craig-plutoflume, can you share more details about the exact patterns you're using?  Are they the exact ones from filed issue?

This fix built upon some additional work that was done in .NET 7, namely the ability to look beyond the next construct in a pattern to determine whether a loop can be made atomic.  Previously we could only ask "can this be made atomic based on just the next construct, yes or no", and now we have three buckets "yes, no, and it depends on what comes next after that".  This fix depends on that third category. \w loops followed by \b are not uncommon, but \w loops followed by a \b followed by something else that overlaps with \w is rare, so this fix enables us to keep the optimization and be correct.  Backporting just this fix to 6.0 would mean giving up the optimization entirely, which makes me nervous, as it could significantly regress real use.  And backporting pieces of the previous .NET 7 improvements to be able to look beyond the next character is riskier.

**Comment by stephentoub** [comment]
Note that as a workaround, you can defeat the faulty auto-atomicity operation by adding an empty positive lookahead after the loop in question, e.g.
```C#
var stringToMatch = "sydney bogota berlin tokyo nairobi denver rio";
Console.WriteLine($"Input: {stringToMatch}");

GetMatches(@"(\b(?!bogo|nai)\w*(?=)\b)\w+", stringToMatch);
GetMatches(@"\w*(?=)\b\w+", stringToMatch);

void GetMatches(string expression, string stringToMatch)
{
    var reg = new System.Text.RegularExpressions.Regex(expression);

    Console.WriteLine($"\nOutput:");
    foreach (var match in reg.Matches(stringToMatch))
    {
        Console.WriteLine(match);
    }
}
```

**Comment by craig-plutoflume** [comment]
Thanks @stephentoub.  For now we've worked around it by compiling the library that does our regex work in `netstandard2` while we investigate the performance of your workaround above.

**Comment by stephentoub** [comment]
> For now we've worked around it by compiling the library that does our regex work in netstandard2

But still running on .NET 6? The surface area you target at build time won't affect this run-time behavior.

**Comment by craig-plutoflume** [comment]
@stephentoub - Thank you for that.  It uncovered an issue in our test harness with the .net6 migration as our test harness showed this approach works, but after you comment we dug a little deeper and saw the error in the harness itself.  You saved us a production incident!

**Pull request merged**